### PR TITLE
Parent child relationships for html publications

### PR DESCRIPTION
**Pull request by peteglondon**
# What 
Process parent/child relationships for `html_publications`

Uses links> parent and links > children to populate parent/child relationships
between publications and html attachments.

Also includes a fix for multi-part documents (guides, travel advice) which was
not working as was. Now refactored to use the same code as `html_publications`.

# Why

We want to display these relationships/related data in `content-data`

Trello: https://trello.com/c/cBmiTJjc/1497-5-add-streams-processing-and-rake-task-to-populate-parent-child-relationships

---
# Review Checklist
* [ ] Changes in scope.
* [ ] Added/updated unit tests.
* [ ] Added/updated feature tests.
* [ ] Added/updated relevant documentation.
* [ ] Added to Trello card.

### CODE REVIEW

**Comment by peteglondon** [comment]
[(5) Add streams processing and rake task to populate parent/child relationships](https://trello.com/c/cBmiTJjc/1497-5-add-streams-processing-and-rake-task-to-populate-parent-child-relationships)

File: app/domain/streams/messages/base_message.rb [hunk 1]
```diff
@@ -32,7 +32,7 @@ def build_attributes(base_path:, title:, document_text:, warehouse_item_id:, sib
       warehouse_item_id: warehouse_item_id,
       withdrawn: withdrawn_notice?,
       historical: historically_political?,
-    }
+    }.merge(Streams::ParentChild::Parser.parse(@payload))
```
[the remark this answers is not in the document]
Comment (peteglondon) [reply app/domain/streams/messages/base_message.rb]: I wanted the adding of the keys to be conditional. I could do it that way and then add `.compact`

File: spec/domain/streams/messages/single_item_message_spec.rb [hunk 2]
```diff
@@ -61,4 +61,35 @@
...
+        'child-2-id:en'
+      ]
+      expect(instance.edition_attributes[:child_sort_order]).to eq(expected)
+    end
+
+    it 'sets the parent warehouse id under temp key' do
+      expect(instance.edition_attributes[:parent_warehouse_id]).to eq('parent-id:en')
+    end
+  end
 end
```
Comment (theseanything) [inline spec/domain/streams/messages/single_item_message_spec.rb]: Can we have test for if the `parent` or `child` key is missing?

Comment (theseanything) [reply spec/domain/streams/messages/single_item_message_spec.rb]: (maybe add them as unit tests for the class)

File: app/domain/streams/handlers/base_handler.rb [hunk 3]
```diff
@@ -12,10 +12,12 @@ def update_editions(items_with_old_editions)
 private
 
   def update_edition(new_edition_attr, old_edition, publishing_api_event)
-    attributes = new_edition_attr.merge(publishing_api_event: publishing_api_event)
+    parent_warehouse_id = new_edition_attr[:parent_warehouse_id]
+    attributes = new_edition_attr.except(:parent_warehouse_id).merge(publishing_api_event: publishing_api_event)
```
Comment (theseanything) [inline app/domain/streams/handlers/base_handler.rb]: Maybe swap these lines around and you can remove the `except`?

File: app/domain/streams/messages/base_message.rb [hunk 4]
```diff
@@ -7,6 +7,11 @@ def initialize(payload, routing_key)
   end
 
   def build_attributes(base_path:, title:, document_text:, warehouse_item_id:, sibling_order: nil)
+    parser = Streams::ParentChild::Parser.new
+    parent_child_attrs = {
+      parent_warehouse_id: parser.get_parent_id(payload),
+      child_sort_order: parser.get_children_ids(payload)
+    }.compact
```
Comment (theseanything) [inline app/domain/streams/messages/base_message.rb]: Why do we need to use compact?

File: app/domain/streams/messages/multipart_message.rb [hunk 5]
```diff
@@ -28,6 +28,9 @@ def edition_attributes
           sibling_order: index
         )
       end
+      attrs.first[:child_sort_order] = attrs.drop(1).map { |h| h[:warehouse_item_id] }
```
Comment (theseanything) [inline app/domain/streams/messages/multipart_message.rb]: Can we assign `attrs.drop(1)` to variable and give it a meaningful name?

File: app/domain/streams/parent_child/links_processor.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      child.update_attributes(parent: @edition, sibling_order: sort_order_for(child.warehouse_item_id))
+    end
+  end
+
+  def sort_order_for(warehouse_item_id)
+    sort_hash[warehouse_item_id] + 1 || 1
+  end
+
+  def sort_hash
+    @sort_hash ||= edition.child_sort_order.map.with_index { |id, index| [id, index] }.to_h
```
[the remark this answers is not in the document]
Comment (peteglondon) [reply app/domain/streams/parent_child/links_processor.rb]: Really? surely this will perform worse and give us no extra value.

**Pull request merged**